### PR TITLE
I/O utilities

### PR DESCRIPTION
**Pull request by oleorhagen**
This adds to new utilities to the `io` library.

1. A `StringReader` to easily wrap a string in a `Reader` interface.
2. A `Discard` writer. Pretty much a `/dev/null` for io Read/Writers.

### CODE REVIEW

File: common/io.hpp [hunk 1]
```diff
@@ -86,6 +88,26 @@ class Discard : virtual public Writer {
...
 };
 
+class StringReader : virtual public Reader {
+private:
+	StreamReader reader_;
+	std::stringstream s_;
+
+public:
+	StringReader(string &str) :
+		s_ {str},
```
Comment (vpodzime) [inline common/io.hpp:98]: Any reason why you prefer the `{}` initializers instead of parentheses? You seem to be the only one using them.

**Review by vpodzime — APPROVED** [review]
> Whenever you want to empty a reader.

Wouldn't it be better to have a way to empty a reader? Having the `Discard` writer makes sense to me, but this use case just suggests wasting CPU cycles.

File: common/io.hpp (hunk 1 above)
Comment (oleorhagen) [reply common/io.hpp:98]: It's just cause it's what I've learnt, and it is what is recommended to use these days appareantly: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-list.

 No hard preference here though.

**Comment by oleorhagen** [comment]
> > Whenever you want to empty a reader.
> 
> Wouldn't it be better to have a way to empty a reader? Having the `Discard` writer makes sense to me, but this use case just suggests wasting CPU cycles.

Mby. It reads pretty nice though I think, and it's a nice drop-in for instance in tests. So in a lot of patterns you will wrap a `Reader` in another `Reader` for instance, and then writer to whatever. But when testing this, you don't necessarily care where you write to. In this case the `Discard` writer is a really nice stand-in.

**Comment by vpodzime** [comment]
> In this case the `Discard` writer is a really nice stand-in.

Yup! Like I said, it makes sense to me. But not to waste CPU cycles outside of tests.

**Pull request merged**